packages/router/utils.ts

```diff
@@ -415,48 +419,98 @@ function flattenRoutes<
-        routesMeta,
-      });
+      for (let exploded of explodeOptionalSegments(route.path)) {
+        flattenRoute(route, index, exploded);
+      }
     }
   });
 
   return branches;
 }
 
+/**
+ * Computes all combinations of optional path segments for a given path.
+ */
+let _explodeOptionalSegments = (path: string): string[] => {
+  let segments = path.split("/");
+  if (segments.length === 0) return [];
+
+  let [first, ...rest] = segments;
+
+  // Optional path segments are denoted by a trailing `?`
+  let isOptional = first.endsWith("?");
+  // Compute the corresponding required segment: `foo?` -> `foo`
+  let required = first.replace(/\?$/, "");
+
+  if (rest.length === 0) {
+    // Intepret empty string as omitting an optional segment
+    // `["one", "", "three"]` corresponds to omitting `:two` from `/one/:two?/three` -> `/one/three`
+    return isOptional ? ["", required] : [required];
+  }
+
+  let restExploded = _explodeOptionalSegments(rest.join("/"));
+  return restExploded.flatMap((subpath) => {
+    // /one + / + :two/three -> /one/:two/three
+    let requiredExploded = subpath === "" ? required : required + "/" + subpath;
+    // For optional segments, return the exploded path _without_ current segment first (`subpath`)
+    // and exploded path _with_ current segment later (`subpath`)
+    // This ensures that exploded paths are emitted in priority order
+    // `/one/three/:four` will come before `/one/three/:five`
+    return isOptional ? [subpath, requiredExploded] : [requiredExploded];
+  });
+};
+
+/**
+ * Computes all combinations of optional path segments for a given path,
+ * excluding combinations that are ambiguous and of lower priority.
+ *
+ * For example, `/one/:two?/three/:four?/:five?` explodes to:
+ * - `/one/three`
+ * - `/one/:two/three`
+ * - `/one/three/:four`
+ * - `/one/:two/three/:four`
+ * - `/one/three/:four/:five`
+ * - `/one/:two/three/:four/:five`
+ *
+ * Note that these paths are not returned:
+ * - `/one/three/:five` (because `/one/three/:four` has priority)
+ * - `/one/:two/three/:five` (because `/one/:two/three/:four` has priority)
+ */
+let explodeOptionalSegments = function* (path: string) {
+  // Compute hash for dynamic path segments
+  // /one/:two/three/:four -> /one/:/three/:
+  let dynamicHash = (subpath: string) =>
+    subpath
+      .split("/")
+      .map((segment) => (segment.startsWith(":") ? ":" : segment))
+      .join("/");
+
+  let dynamicHashes = new Set<string>();
+  for (let exploded of _explodeOptionalSegments(path)) {
+    let hash = dynamicHash(exploded);
+
+    // `/one/three/:four` and `/one/three/:five` have the same hash: `/one/three/:`
+    // so we only emit the first one of these we come across
+    // _explodeOptionalSegments returns exploded paths in priority order,
+    // so `/one/three/:four` will come before `/one/three/:five`
+    if (dynamicHashes.has(hash)) continue;
+
+    dynamicHashes.add(hash);
+    yield exploded;
```

Comment:
Is the generator functionally required here?  Or could we just return an array of exploded paths to loop through?

Reply:
Its equivalent to:

```ts
let explodedPaths = []
for (...) {
   ...
   explodedPaths.push(exploded)
}
return exploded
```

I just think its easier to read and write these sorts of functions as generators since there is less bookkeeping to do, plus its lazily evaluated in case you don't need the full set (e.g. if you return early after one match or something).

Not a hill I want to die on though, so if you feel strongly I'm ok to switch this to a non-generator function.

Reply:
Yeah I think if we were returning early it could make more sense (so there may be a use case in the eventual trie walk), but if we're processing every entry I find it's harder to debug since I can't just look at the output of `explodeOptionalSegments` in isolation.  I think I'd want a particularly compelling use case to introduce the first generator to the codebase 😂